### PR TITLE
Add UnixSocketPort

### PR DESCRIPTION
**Pull request by kot-begemot-uk**
This adds UnixSocketPort functionality.

Changes to the "Switch" which include a default "monitor" port to follow

### CODE REVIEW

File: pkg/network/modules.go [hunk 1]
```diff
@@ -169,3 +169,9 @@ func (m *Module) SetIGateMap(gMap map[Gate]PipelineModule) {
 func (m *Module) SetEGateMap(gMap map[Gate]PipelineModule) {
 	m.IGates = gMap
 }
+
+func MakeGateMap() map[Gate]PipelineModule {
+    rGates := make(map[Gate]PipelineModule)
+    rGates[0] = nil
```
Comment (trozet) [inline pkg/network/modules.go]: will this work for ingress gates with bess? I know a lot bess modules only have a single ingress gate (0) for their module.

File: pkg/network/modules.go [hunk 2]
```diff
@@ -63,12 +63,14 @@ type Port struct {
...
 	Virtual    bool
 	DPDK       bool
+        UnixSocket bool
 	IfaceName  string
 	Network    string
 	NamesSpace string
 	IPAddr     string
 	MacAddr    string
 	Gateway    string
+        UnixPath   string //used in Socket Port and DPDK vhost-user
```
Comment (trozet) [inline pkg/network/modules.go]: In the DPDK patch I added SocketPath field:
https://github.com/nimbess/nimbess-agent/pull/11/commits/5fa5e7c1860723363198c3fbd73fec900cde4286#diff-4cee0960744d01b5f4d7354d8b48b8b0R72

Can you please  these earlier commits on top of that and re-use that var?

Comment (kot-begemot-uk) [reply pkg/network/modules.go]: It is reused in the latest version - note "outdated" on the github interface.

File: pkg/network/modules.go (hunk 1 above)
Comment (kot-begemot-uk) [reply pkg/network/modules.go]: It can be zero actually. There are modules with no iGates.
Ditto for eGates.
Bess gets around this by having two constants for each module which are fed into the registration macro.

Comment (kot-begemot-uk) [reply pkg/network/modules.go]: So the right way of doing it should be having MaxIGates and MaxEGates as part of the struct and building the map accordingly.

Comment (kot-begemot-uk) [reply pkg/network/modules.go]: The only place where IGates are used are replicator and forwarder. 
In the replicator gate 0 is immediately set to the forwarder after being added in the creation function.
In the l2forward case it should not work anyway as it is single ingress gate in Bess +/- my patches and the go code assumes it is multiple - I will have to fix that to get a pipeline per port as intended.

Comment (trozet) [reply pkg/network/modules.go]: We need to be careful about separating bess specific stuff from abstract nimbess stuff. What I mean is, just because a BESS l2 fwd module only allows 1 ingress gate, doesn't necessarily mean a nimbess switch should. I think it is something we should discuss in the future.

Comment (trozet) [reply pkg/network/modules.go]: The Ingress and Egress Port inherit module and have igates,egates. The egress port will connect its next available ingress gate to the last module of the pipeline. Just need to test if this works if it is non-zero with your changes. I'll try it out.

File: pkg/drivers/bess/bess.go [hunk 3]
```diff
@@ -543,45 +545,67 @@ func (d *Driver) createPort(port *network.Port) error {
...
-		portArg := &bess_pb.PCAPPortArg{Dev: port.IfaceName}
-		portAny, err = ptypes.MarshalAny(portArg)
-		if err != nil {
-			log.Errorf("Failure to serialize kernel port args: %v", portArg)
-			return errors.New("failed to serialize port args")
+		if port.DPDK {
+			return errors.New("DPDK physical ports are not currently supported")
 		}
+		// Must be a kernel iface, use PCAP type port
+                if port.UnixSocket {
```
Comment (trozet) [inline pkg/drivers/bess/bess.go]: this should be under the virtual type port if statement right?

**Comment by kot-begemot-uk** [comment]
I think I addressed the NPE, it was originating from further the two gate enumeration loops (second line in backtrace).

File: pkg/drivers/bess/bess.go (hunk 3 above)
Comment (kot-begemot-uk) [reply pkg/drivers/bess/bess.go]: Comment is wrong. 
Did not move the pcap comment after adding UnixSocket.
It is not virtual, so why should it be under virtual?

File: pkg/network/modules.go (hunk 1 above)
Comment (kot-begemot-uk) [reply pkg/network/modules.go]: I think I fixed the NPE, can you retest please?

**Comment by trozet** [comment]
There's another NPE here:
time="2019-08-26T12:56:21Z" level=debug msg="Building Nimbess switch for: {Module:{Meta:Switch_k8s_pod_network-fastpath Name:Switch_k8s_pod_network-fastpath_port_f123fd1c5bddc2db70c22cc3e78566814f6ca7048416e297973c20634981ac1b_eth0-fastpath IGates:map[0:<nil> 1:0xc000397a40] EGates:map[0:<nil>]} L2FIB:map[]}" func="github.com/nimbess/nimbess-agent/pkg/drivers/bess.(*Driver).createSwitch" file="/go/src/github.com/nimbess/nimbess-agent/pkg/drivers/bess/bess.go:215"
time="2019-08-26T12:56:21Z" level=info msg="Module created: Switch_k8s_pod_network-fastpath_port_f123fd1c5bddc2db70c22cc3e78566814f6ca7048416e297973c20634981ac1b_eth0-fastpath_l2forward, L2Forward" func="github.com/nimbess/nimbess-agent/pkg/drivers/bess.(*Driver).createModule" file="/go/src/github.com/nimbess/nimbess-agent/pkg/drivers/bess/bess.go:322"
time="2019-08-26T12:56:21Z" level=info msg="Module created: Switch_k8s_pod_network-fastpath_port_f123fd1c5bddc2db70c22cc3e78566814f6ca7048416e297973c20634981ac1b_eth0-fastpath_replicate, Replicate" func="github.com/nimbess/nimbess-agent/pkg/drivers/bess.(*Driver).createModule" file="/go/src/github.com/nimbess/nimbess-agent/pkg/drivers/bess/bess.go:322"
time="2019-08-26T12:56:21Z" level=debug msg="Wiring up module: &{Switch:{Module:{Meta: Name:Switch_k8s_pod_network-fastpath_port_f123fd1c5bddc2db70c22cc3e78566814f6ca7048416e297973c20634981ac1b_eth0-fastpath_l2forward IGates:map[0:<nil> 1:0xc000397a40] EGates:map[0:0xc000044200]} L2FIB:map[]}}" func="github.com/nimbess/nimbess-agent/pkg/drivers/bess.(*Driver).wireModule" file="/go/src/github.com/nimbess/nimbess-agent/pkg/drivers/bess/bess.go:330"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x112e238]

goroutine 61 [running]:
github.com/nimbess/nimbess-agent/pkg/drivers/bess.(*Driver).wireModule(0xc000264d20, 0x157eb00, 0xc000386c80, 0xc000366580, 0x7d)
        /go/src/github.com/nimbess/nimbess-agent/pkg/drivers/bess/bess.go:334 +0x488
github.com/nimbess/nimbess-agent/pkg/drivers/bess.(*Driver).createSwitch(0xc000264d20, 0xc000397840, 0x1358120, 0x1)
        /go/src/github.com/nimbess/nimbess-agent/pkg/drivers/bess/bess.go:242 +0x4ea
github.com/nimbess/nimbess-agent/pkg/drivers/bess.(*Driver).RenderModules(0xc000264d20, 0xc0004e46e0, 0x2, 0x2, 0xc0000accc0, 0x0)
        /go/src/github.com/nimbess/nimbess-agent/pkg/drivers/bess/bess.go:190 +0x67d
github.com/nimbess/nimbess-agent/pkg/agent.(*NimbessAgent).createPortPipeline(0xc0003630e0, 0xc000357560, 0xc0000e9720, 0x18, 0x0, 0x0, 0x0)
        /go/src/github.com/nimbess/nimbess-agent/pkg/agent/agent.go:124 +0x652
github.com/nimbess/nimbess-agent/pkg/agent.(*NimbessAgent).Add(0xc0003630e0, 0x156d280, 0xc00023d710, 0xc000357200, 0x0, 0x0, 0x0)
        /go/src/github.com/nimbess/nimbess-agent/pkg/agent/agent.go:336 +0x8ba
github.com/nimbess/nimbess-agent/pkg/proto/cni._RemoteCNI_Add_Handler(0x131d6e0, 0xc0003630e0, 0x156d280, 0xc00023d710, 0xc000265800, 0x0, 0x0, 0x0, 0xc00017e620, 0x6d)
        /go/src/github.com/nimbess/nimbess-agent/pkg/proto/cni/cni.pb.go:504 +0x23e
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00022b8c0, 0x1579980, 0xc00022c180, 0xc00023a300, 0xc000361c50, 0x1fe1060, 0x0, 0x0, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:995 +0x485
google.golang.org/grpc.(*Server).handleStream(0xc00022b8c0, 0x1579980, 0xc00022c180, 0xc00023a300, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:1275 +0xe02
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0004dc1d0, 0xc00022b8c0, 0x1579980, 0xc00022c180, 0xc00023a300)
        /go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:710 +0x9f
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:708 +0xa1

**Comment by kot-begemot-uk** [comment]
> There's another NPE here:

I think I addressed it. Looking through the code once more if there are any left.

I definitely will need to sort out my k8 testbed though - I have only a proper BESS testbed at the moment without the control plane on top of it.

**Comment by kot-begemot-uk** [comment]
While at it, the logic in the wiring is wrong - it assumes that a gate is always wired module to module. That is not the case - it may be the special gate value DROPGATE.

File: pkg/drivers/bess/bess.go [hunk 4]
```diff
@@ -38,11 +38,14 @@ const (
...
 	MODULE    = "Module"
 	VPORT     = "VPort"
+	PMDPORT   = "PMDPort"
 	PCAPPORT  = "PCAPPort"
+	UNIXPORT  = "UnixSocketPort"
 	PORTOUT   = "PortOut"
 	PORTINC   = "PortInc"
 	L2FORWARD = "L2Forward"
 	REPLICATE = "Replicate"
+    DROPGATE  = 8192    // This needs to be adjusted if MAX_GATES in bess changes
```
Comment (trozet) [inline pkg/drivers/bess/bess.go]: Can you explain what this is used for?

File: pkg/agent/pipeline.go [hunk 5]
```diff
@@ -148,21 +148,25 @@ func (s *NimbessPipeline) createPipeline(module network.PipelineModule, port str
...
+		newMod.(network.PipelineModule).SetIGateMap(network.MakeGateMap())
 		// Append new module to existing module list
 		s.Modules = append(s.Modules, newMod.(network.PipelineModule))
 		log.Debugf("Module created: %+v", newMod)
 	}
 
 	// Walk meta module Ingress gates and connect new port pipeline modules
 	for gate, mod := range module.GetIGateMap() {
-		newMod.Connect(s.createPipeline(mod, port), false, &gate)
+		if mod != nil {
```
Comment (trozet) [inline pkg/agent/pipeline.go]: I'm a little wary of this behavior where it is ok that a module is nil. The reason why is that it will be hard to differentiate if we are intentionally having a mod as nil (reserved gate) or if there is a real bug in the code. The previous way the code worked is that there shouldn't be any nil pointer. I'm thinking it might be better to make a new Nimbess Module type called Reserved and use that to represent an intentional act to reserve the gate. Then in all these places we check to see if mod == reflect.TypeOf(&modules.Reserved{}). I think it is still also good to keep the checks for nil, and raise an error if a mod is nil. Will help debug issues in the future.

Comment (kot-begemot-uk) [reply pkg/agent/pipeline.go]: I am reaching the same conclusion.

However instead of Reserved, I am thinking to use here a sink instance to implement a default "drop if you do not know what you are doing"

File: pkg/drivers/bess/bess.go (hunk 4 above)
Comment (kot-begemot-uk) [reply pkg/drivers/bess/bess.go]: In BESS code it is "drop"

However, I think it is cheaper to actually wire to a SINK instead. See my other comment.

File: pkg/agent/agent.go [hunk 6]
```diff
@@ -74,6 +74,13 @@ var vhostIndex = 0
 // It contains a loadable runtime data plane driver to manage the data plane.
```
Comment (trozet) [inline pkg/agent/agent.go]: these comments should be on top of type NimbessAgent (line 84)

File: pkg/agent/agent.go [hunk 7]
```diff
@@ -85,6 +92,8 @@ type NimbessAgent struct {
 	Pipelines map[string]*NimbessPipeline
 	// MetaPipelines contains a map of network name to abstract pipelines
 	MetaPipelines map[string]*NimbessPipeline
+    l2fib map[string]L2FIBEntry
```
Comment (trozet) [inline pkg/agent/agent.go]: something wrong with the spacing here, can you please run:
gofmt -w -s pkg/* && gofmt -w -s cmd/*

Comment (trozet) [reply pkg/agent/agent.go]: In fact what we could do is make the Meta pipeline switch module hold the L2FIB (it already has a map for that), then all of the port pipelines created from this Meta pipeline could set their L2FIB as a pointer to the Meta Switch L2FIB. Then you only need to modify the FIB of the meta switch.

File: pkg/agent/agent.go [hunk 8]
```diff
@@ -628,6 +675,60 @@ func annotatePod(k8sClient kubernetes.Interface, req *cni.CNIRequest, configData
...
+    defer s.Mu.Unlock()
+
+    if fibRequest.Command == "LEARN" {
+        // Learned
+        s.l2fib[fibRequest.MAC] = L2FIBEntry {
+            permanent: false,
+            age:fibRequest.Setage,
+            port:fibRequest.Port,
+        }
+        for key, pipeline := range s.Pipelines {
```
[the remark this answers is not in the document]
Comment (trozet) [reply pkg/agent/agent.go]: or use my idea above of just updating the meta pipeline, FIB, then triggering update to all affected port pipelines

File: pkg/drivers/bess/bess.go [hunk 9]
```diff
@@ -255,13 +291,34 @@ func (d *Driver) createSwitch(module *network.Switch) error {
 	// Egress gates from original Switch module will be used for L2Forward lookup match, and then
 	// also used for replicator flooding. Add Egress Gate for L2Forward -> replicator and Ingress
 	// gate for reverse
-	l2Fwd.EGates[999] = rep
+	l2Fwd.EGates[0] = rep
```
Comment (trozet) [inline pkg/drivers/bess/bess.go]: I think this works because the Render is called in agent.go before egress ports are added to the pipeline. That way we know that no egress gates on this switch have been reserved yet. I think it would be good to add a comment here stating this to avoid issues in the future.

File: pkg/agent/agent.go [hunk 10]
```diff
@@ -85,6 +92,8 @@ type NimbessAgent struct {
 	Pipelines map[string]*NimbessPipeline
 	// MetaPipelines contains a map of network name to abstract pipelines
 	MetaPipelines map[string]*NimbessPipeline
+    l2fib map[string]L2FIBEntry
+    notifications chan network.L2FIBCommand
```
Comment (trozet) [inline pkg/agent/agent.go]: in the future, we may want to make this chan of a more generic type. Then we can process multiple different types of notifications on the same channel. What do you think?

Comment (kot-begemot-uk) [reply pkg/agent/agent.go]: That means using reflection.*. I tried that, the recovery a complex notification via reflection is beyond ugly.

File: pkg/drivers/bess/bess.go (hunk 9 above)
Comment (kot-begemot-uk) [reply pkg/drivers/bess/bess.go]: Ack.

File: pkg/agent/agent.go [hunk 11]
```diff
@@ -410,6 +422,38 @@ func (s *NimbessAgent) updateL2FIB(l2Forwarder *network.Switch, macAddr string,
 	return nil
 }
 
+func (s *NimbessAgent) addMACtoL2FIB(MAC string, pipeline *NimbessPipeline, portname string) error {
```
[the remark this answers is not in the document]
Comment (kot-begemot-uk) [reply pkg/agent/agent.go]: Ack and ack.

File: pkg/agent/agent.go (hunk 7 above)
Comment (kot-begemot-uk) [reply pkg/agent/agent.go]: Looks like you are using hard tabs while I have mine set-up for soft-tab which the editor fills with spaces after that. I use hard-tap only for the linux source where it is a code style requirement (no spaces for indentation).

Comment (kot-begemot-uk) [reply pkg/agent/agent.go]: > In fact what we could do is make the Meta pipeline switch module hold the L2FIB (it already has a map for that), then all of the port pipelines created from this Meta pipeline could set their L2FIB as a pointer to the Meta Switch L2FIB. Then you only need to modify the FIB of the meta switch.

I need to think on this as this will require reorganizing quite a few things - f.e. we may need to spin up an l2fib notification thread per metapipeline instead of a centralized agent thread and manage notification channels accordingly. 

For now, I will limit the mac fib updates so that they are limited to the same metakey - this takes care of network isolation.

File: pkg/agent/agent.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (kot-begemot-uk) [reply pkg/agent/agent.go]: Done by checking the key. We need to revisit all the data structures after ONS.

It supports static macs at present by the way, you do not need to skip them if learning is enabled.

File: pkg/agent/pipeline.go [hunk 12]
```diff
@@ -151,12 +151,16 @@ func (s *NimbessPipeline) createPipeline(module network.PipelineModule, port str
...
+            newMod.Connect(s.createPipeline(mod, port), false, &gate)
+		}
 	}
 
 	// Walk meta module Egress gates and connect new port pipeline modules
 	for gate, mod := range module.GetEGateMap() {
-		newMod.Connect(s.createPipeline(mod, port), true, &gate)
+		if mod != nil {
```
[the remark this answers is not in the document]
Comment (kot-begemot-uk) [reply pkg/agent/pipeline.go]: Warning - yes, error not in all places as there is  no way to return it up for some of the use cases.

File: pkg/drivers/bess/reader.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+        if (err == nil) {
+            r.c = c
+            return r
+        }
+        if err != nil {
+            log.Debugf("Could Not connect to %s, error %s, retry no %d in one second:", r.path, err, i)
+        }
+        time.Sleep(time.Second)
+    }
+    log.Fatal("Could Not connect:")
```
Comment (trozet) [inline pkg/drivers/bess/reader.go]: Please change this to log.Error and return a nil reader or error to the calling function and handle it there. This is important so we handle the CNI reply gracefully that it failed and not just exit the program

File: pkg/agent/agent.go [hunk 14]
```diff
@@ -628,6 +687,74 @@ func annotatePod(k8sClient kubernetes.Interface, req *cni.CNIRequest, configData
...
+		log.Debugf("Processing FIB Request %s MAC %s", fibRequest.Command, fibRequest.MAC)
+
+	s.Mu.Lock()
+	defer s.Mu.Unlock()
+
+	// pipeline's fib points to its parent metapipeline fib
+	pipeline, ok := s.Pipelines[fibRequest.Port]
+
+	if (!ok) {
+		log.Errorf("Failed to find metaPipeline and its fib")
```
Comment (trozet) [inline pkg/agent/agent.go]: are you missing a return here if cannot find metapipeline/fib? Otherwise I get:

time="2019-10-17T13:14:14Z" level=debug msg="Processing FIB Request LEARN MAC 52:71:42:3c:f9:01" func="github.com/nimbess/nimbess-agent/pkg/agent.(*NimbessAgent).processNotification" file="/go/src/github.com/nimbess/nimbess-agent/pkg/agent/agent.go:693"
time="2019-10-17T13:14:14Z" level=error msg="Failed to find metaPipeline and its fib" func="github.com/nimbess/nimbess-agent/pkg/agent.(*NimbessAgent).processNotification" file="/go/src/github.com/nimbess/nimbess-agent/pkg/agent/agent.go:702"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0x13def10]

goroutine 27 [running]:
github.com/nimbess/nimbess-agent/pkg/agent.(*NimbessAgent).processNotification(0xc00047c200, 0x0)
	/go/src/github.com/nimbess/nimbess-agent/pkg/agent/agent.go:707 +0x5b0
github.com/nimbess/nimbess-agent/pkg/agent.(*NimbessAgent).runNotifications(0xc00047c200)
	/go/src/github.com/nimbess/nimbess-agent/pkg/agent/agent.go:752 +0x59
created by github.com/nimbess/nimbess-agent/pkg/agent.(*NimbessAgent).Run
	/go/src/github.com/nimbess/nimbess-agent/pkg/agent/agent.go:815 +0x17c

File: pkg/drivers/bess/reader.go (hunk 13 above)
Comment (trozet) [reply pkg/drivers/bess/reader.go]: also need to add the unix socket path (shared-dir) to the mount for the agent in nimbess.yaml

File: pkg/drivers/bess/bess.go [hunk 15]
```diff
@@ -348,10 +348,10 @@ func (d *Driver) createSwitch(module *network.Switch) error {
 		log.Errorf("Failed to set default gate for l2forward, error: %s", res.GetError().GetErrmsg())
 		return errors.New(res.GetError().GetErrmsg())
 	}
-    // We need to recover the original mod name from Switch_%s
-    orig_name := module.GetName() //expected as Switch_%s - from pipeline module
-    r := NewReader(orig_name[len("Switch_"):], port.SocketPath, d.notifications)
```
Comment (trozet) [inline pkg/drivers/bess/bess.go]: There is still an issue here where this port name isn't right. The format of the port name will be "Switch_<meta network>_<port name>". An easier way to remove the meta portion is to just call module.GetMeta(). Here is the diff for a patch I used and it seems to be working (I see macs learned and no errors). Can you please update this part and then I think we are good to merge!
```
[trozet@trozet nimbess-agent]$ git diff
diff --git a/pkg/agent/agent.go b/pkg/agent/agent.go
index af17316..a394f57 100644
--- a/pkg/agent/agent.go
+++ b/pkg/agent/agent.go
@@ -699,7 +699,7 @@ func (s *NimbessAgent) processNotification() (bool) {
        pipeline, ok := s.Pipelines[fibRequest.Port]
 
        if (!ok) {
-               log.Errorf("Failed to find metaPipeline and its fib")
+               log.Errorf("Failed to find metaPipeline and its fib for port: %s", fibRequest.Port)
                return true /* This is bad, but it is not a reason to exit the notification thread */
        }
 
diff --git a/pkg/drivers/bess/bess.go b/pkg/drivers/bess/bess.go
index 2497933..beda457 100644
--- a/pkg/drivers/bess/bess.go
+++ b/pkg/drivers/bess/bess.go
@@ -349,8 +349,14 @@ func (d *Driver) createSwitch(module *network.Switch) error {
                return errors.New(res.GetError().GetErrmsg())
        }
        // We need to recover the original mod name from Switch_%s
-       orig_name := module.GetName() //expected as Switch_%s - from pipeline module
-       r := NewReader(orig_name[len("Switch_"):], port.SocketPath, d.notifications)
+       origName := module.GetName()
+       metaName := module.GetMeta() //expected as Switch_<meta name> - from pipeline module
+       if len(metaName) == 0 {
+               log.Warningf("Meta pipeline empty for Switch: %s", origName)
+       }
+       derivedPortName := origName[len(metaName)+1:] // remove "<MetaName>_" prefix
+       log.Debugf("Module name: %s, meta name: %s, port name: %s", origName, metaName, derivedPortName)
+       r := NewReader(derivedPortName, port.SocketPath, d.notifications)
        go r.Run()
        log.Infof("BESS Switch created with L2FWD: %s, Replicate: %s", l2Fwd.Name, rep.Name)
        return nil
```

Comment (trozet) [reply pkg/drivers/bess/bess.go]: Sorry just realized github parsed my 2nd sentence and it is wrong in the comment. The format of the port name will be '''Switch_<meta network>'''

Comment (trozet) [reply pkg/drivers/bess/bess.go]: Still parsing it haha so will just state it like this Switch_metaNetwork
